feature: Configurable logging channels per log level in Settings

## Summary

Добавить в Settings Tab плагина матрицу маршрутизации логов: пользователь выбирает, в какие каналы транслируется каждый уровень логирования. Три канала: **Notice** (Obsidian UI), **Console** (developer tools), **File** (`exocortex-logs.txt` в корне vault). File-канал — новая функциональность.

## Motivation

Сейчас Logger жёстко пишет в console, а NotificationService — в Notice + console. Пользователь не может:
- Отключить шумные Notice для debug/info
- Включить file logging для диагностики без открытия DevTools
- Настроить routing per level (например: error → все 3 канала, debug → только file)

## Acceptance Criteria (BDD)

### Settings UI
- [ ] **Given** user opens Settings Tab **When** scrolls to "Logging" section **Then** sees a matrix: rows = `debug | info | warn | error`, columns = `Notice | Console | File`, each cell is a toggle
- [ ] **Given** user toggles a channel for a level **When** saves settings **Then** preference persists across plugin reload

### Default Configuration
- [ ] **Given** fresh install **When** settings loaded **Then** defaults are:
  | Level | Notice | Console | File |
  |-------|--------|---------|------|
  | debug | ❌ | ✅ | ❌ |
  | info  | ❌ | ✅ | ❌ |
  | warn  | ✅ | ✅ | ❌ |
  | error | ✅ | ✅ | ❌ |

### Console Channel
- [ ] **Given** Console enabled for a level **When** log emitted **Then** output appears in DevTools console (existing behavior preserved)

### Notice Channel
- [ ] **Given** Notice enabled for `info` **When** info log emitted **Then** Obsidian Notice shown with appropriate icon
- [ ] **Given** Notice disabled for `info` **When** info log emitted **Then** no Notice appears

### File Channel (NEW)
- [ ] **Given** File enabled for any level **When** log emitted **Then** line appended to `<vault>/exocortex-logs.txt`
- [ ] **Given** File enabled **When** plugin starts **Then** file created if not exists (no error if already exists)
- [ ] **Given** log line written to file **Then** format is `[ISO8601] [LEVEL] [Context] message`
- [ ] **Given** File disabled for all levels **When** plugin runs **Then** no file I/O occurs, file not created

## Technical Notes

### Affected Files
- `src/domain/settings/ExocortexSettings.ts` — add `logChannels` to settings interface
- `src/presentation/settings/ExocortexSettingTab.ts` — add logging matrix UI
- `src/adapters/logging/Logger.ts` — route logs through channel config
- `src/adapters/logging/ILogger.ts` — extend if needed for channel awareness
- `src/adapters/logging/LoggerFactory.ts` — inject settings into Logger instances
- NEW: `src/adapters/logging/FileLogChannel.ts` — file appender using Obsidian Vault API

### Settings Model
```typescript
interface LogChannelConfig {
  notice: boolean;
  console: boolean;
  file: boolean;
}

interface ExocortexSettings {
  // ... existing fields
  logChannels: {
    debug: LogChannelConfig;
    info: LogChannelConfig;
    warn: LogChannelConfig;
    error: LogChannelConfig;
  };
}
```

### Architecture Approach
1. **Logger** получает `logChannels` config через DI (settings injection)
2. Каждый `logger.info()` / `.error()` etc. проверяет config и роутит в enabled каналы
3. **FileLogChannel** использует `vault.adapter.append()` для записи в файл (Obsidian API, не Node fs)
4. **NotificationService** остаётся для user-facing уведомлений, но Logger тоже может роутить в Notice channel для system-level логов

### File Format
```
[2026-04-11T10:30:45.123+05:00] [INFO] [SPARQLEngine] Query executed: 42 results in 150ms
[2026-04-11T10:30:46.456+05:00] [ERROR] [VaultStore] Failed to read file: abc123.md
```

## AI Guidance

### Entry Points
1. Start with `ExocortexSettings.ts` — add `logChannels` field + defaults
2. Create `FileLogChannel.ts` — vault-based file appender
3. Modify `Logger.ts` — add channel routing logic
4. Update `ExocortexSettingTab.ts` — add matrix UI
5. Wire through DI in plugin bootstrap

### Patterns to Follow
- Settings pattern: see existing toggles in `ExocortexSettingTab.ts` (e.g. `showEffortArea`)
- DI pattern: see `ObsidianNotificationService` registration
- Vault file ops: use `this.app.vault.adapter.append()` / `.exists()` / `.write()`

### Constraints
- Do NOT use Node.js `fs` — use Obsidian Vault API for mobile compatibility
- File channel must be lazy — no file I/O if all file toggles are off
- Preserve existing Logger API — `logger.info()` signature unchanged

## Size
- [ ] S (≤4h) / [x] M (4-16h) / [ ] L (16+h)

--- a/packages/obsidian-plugin/src/ExocortexPlugin.ts
+++ b/packages/obsidian-plugin/src/ExocortexPlugin.ts
@@ -9,10 +9,13 @@ import { container } from "tsyringe";
 import { UniversalLayoutRenderer } from "./presentation/renderers/UniversalLayoutRenderer";
 import { ILogger } from "./adapters/logging/ILogger";
 import { LoggerFactory } from "./adapters/logging/LoggerFactory";
+import { Logger } from "./adapters/logging/Logger";
+import { FileLogChannel } from "./adapters/logging/FileLogChannel";
 import { CommandManager } from "./application/services/CommandManager";
 import {
   ExocortexSettings,
   DEFAULT_SETTINGS,
+  DEFAULT_LOG_CHANNELS,
 } from "./domain/settings/ExocortexSettings";
 import { ExocortexSettingTab } from "./presentation/settings/ExocortexSettingTab";
 import {
@@ -83,6 +86,7 @@ export default class ExocortexPlugin extends Plugin {
   private bodyLinkPatch!: BodyLinkPatch;
   private propertiesUidCopyPatch!: PropertiesUidCopyPatch;
   private graphViewPatch!: GraphViewPatch;
+  private fileLogChannel!: FileLogChannel;
 
   override async onload(): Promise<void> {
     try {
@@ -96,6 +100,10 @@ export default class ExocortexPlugin extends Plugin {
       this.timerManager = new TimerManager();
 
       await this.loadSettings();
+
+      // Initialize log channel routing from settings
+      this.fileLogChannel = new FileLogChannel(this.app.vault.adapter);
+      this.configureLogChannels();
 
       this.printNameRuleService = new PrintNameRuleService(this.app);
       this.printNameRuleService.initialize();
@@ -387,11 +395,31 @@ export default class ExocortexPlugin extends Plugin {
       this.graphViewPatch.cleanup();
     }
 
+    // Reset log channel routing
+    Logger.resetChannels();
+
     this.logger?.info("Exocortex Plugin unloaded");
   }
 
   async loadSettings(): Promise<void> {
     this.settings = Object.assign({}, DEFAULT_SETTINGS, await this.loadData());
+    // Ensure logChannels exists for users upgrading from older versions
+    if (!this.settings.logChannels) {
+      this.settings.logChannels = DEFAULT_LOG_CHANNELS;
+    }
+  }
+
+  /**
+   * Apply current log channel settings to the Logger subsystem.
+   * Called on init and whenever log channel settings change.
+   */
+  configureLogChannels(): void {
+    const hasFileEnabled = Object.values(this.settings.logChannels).some(c => c.file);
+    Logger.configure({
+      channels: this.settings.logChannels,
+      fileChannel: hasFileEnabled ? this.fileLogChannel : null,
+      noticeCallback: (msg: string) => new Notice(msg),
+    });
   }
 
   async saveSettings(): Promise<void> {

--- a/packages/obsidian-plugin/src/ExocortexPlugin.ts
+++ b/packages/obsidian-plugin/src/ExocortexPlugin.ts
@@ -87,6 +87,7 @@ export default class ExocortexPlugin extends Plugin {
   private propertiesUidCopyPatch!: PropertiesUidCopyPatch;
   private graphViewPatch!: GraphViewPatch;
   private fileLogChannel!: FileLogChannel;
+  private notifier!: ObsidianNotificationService;
 
   override async onload(): Promise<void> {
     try {
@@ -101,7 +102,8 @@ export default class ExocortexPlugin extends Plugin {
 
       await this.loadSettings();
 
-      // Initialize log channel routing from settings
+      // Initialize notification service and log channel routing
+      this.notifier = new ObsidianNotificationService();
       this.fileLogChannel = new FileLogChannel(this.app.vault.adapter);
       this.configureLogChannels();
 
@@ -120,7 +122,7 @@ export default class ExocortexPlugin extends Plugin {
         this.app,
       );
 
-      const notifier = new ObsidianNotificationService();
+      const notifier = this.notifier;
       this.sparqlProcessor = new SPARQLCodeBlockProcessor(this, notifier);
       this.layoutProcessor = new LayoutCodeBlockProcessor(this);
       this.sparql = new SPARQLApi(this);
@@ -418,7 +420,7 @@ export default class ExocortexPlugin extends Plugin {
     Logger.configure({
       channels: this.settings.logChannels,
       fileChannel: hasFileEnabled ? this.fileLogChannel : null,
-      noticeCallback: (msg: string) => new Notice(msg),
+      noticeCallback: (msg: string) => this.notifier.info(msg),
     });
   }
 

--- a/packages/obsidian-plugin/src/adapters/logging/FileLogChannel.ts
+++ b/packages/obsidian-plugin/src/adapters/logging/FileLogChannel.ts
@@ -1,0 +1,42 @@
+import type { DataAdapter } from "obsidian";
+
+const LOG_FILE_NAME = "exocortex-logs.txt";
+
+/**
+ * Appends log lines to a file in the vault root.
+ * Uses Obsidian's DataAdapter (vault.adapter) for mobile compatibility.
+ */
+export class FileLogChannel {
+  private buffer: string[] = [];
+  private flushScheduled = false;
+
+  constructor(private readonly adapter: DataAdapter) {}
+
+  append(level: string, context: string, message: string): void {
+    const timestamp = new Date().toISOString();
+    const line = `[${timestamp}] [${level.toUpperCase()}] [${context}] ${message}\n`;
+    this.buffer.push(line);
+    this.scheduleFlush();
+  }
+
+  private scheduleFlush(): void {
+    if (this.flushScheduled) return;
+    this.flushScheduled = true;
+    // Microtask batching — collects multiple log calls in the same tick
+    queueMicrotask(() => this.flush());
+  }
+
+  private flush(): void {
+    const lines = this.buffer.join("");
+    this.buffer = [];
+    this.flushScheduled = false;
+    if (!lines) return;
+    // Fire-and-forget — logging should never block the caller
+    void this.adapter.append(LOG_FILE_NAME, lines).catch(() => {
+      // If file doesn't exist yet, create it then append
+      void this.adapter.write(LOG_FILE_NAME, lines).catch(() => {
+        // Silently fail — we can't log a logging failure
+      });
+    });
+  }
+}

--- a/packages/obsidian-plugin/src/adapters/logging/Logger.ts
+++ b/packages/obsidian-plugin/src/adapters/logging/Logger.ts
@@ -1,20 +1,27 @@
 /* eslint-disable no-console */
 import type { ILogger, ErrorLogOptions } from "./ILogger";
 import { ErrorMessages } from "./ErrorCodes";
+import type { LogChannelsSettings, LogLevel } from "@plugin/domain/settings/ExocortexSettings";
+import type { FileLogChannel } from "./FileLogChannel";
 
 /**
- * Environment-aware logger that sanitizes stack traces in production.
+ * Callback type for showing Obsidian Notice UI notifications.
+ */
+export type NoticeCallback = (message: string) => void;
+
+/**
+ * Environment-aware logger with configurable channel routing.
  *
- * In production mode:
- * - Shows user-friendly error messages with error codes
- * - Hides stack traces to prevent information leakage
- *
- * In development mode:
- * - Shows full error details including stack traces
- * - Shows additional context for debugging
+ * Each log level can independently route to three channels:
+ * - Console (developer tools)
+ * - Notice (Obsidian UI notification)
+ * - File (exocortex-logs.txt in vault root)
  */
 export class Logger implements ILogger {
   private static isDevelopment: boolean | undefined = undefined;
+  private static channelConfig: LogChannelsSettings | null = null;
+  private static fileChannel: FileLogChannel | null = null;
+  private static noticeCallback: NoticeCallback | null = null;
 
   constructor(private context: string) {}
 
@@ -55,16 +62,67 @@ export class Logger implements ILogger {
     return Logger.checkIsDevelopment();
   }
 
+  /**
+   * Configure channel routing for all Logger instances.
+   * Called once during plugin initialization and whenever settings change.
+   */
+  static configure(options: {
+    channels: LogChannelsSettings;
+    fileChannel?: FileLogChannel | null;
+    noticeCallback?: NoticeCallback | null;
+  }): void {
+    Logger.channelConfig = options.channels;
+    Logger.fileChannel = options.fileChannel ?? null;
+    Logger.noticeCallback = options.noticeCallback ?? null;
+  }
+
+  /**
+   * Reset channel configuration (used in tests and plugin unload).
+   */
+  static resetChannels(): void {
+    Logger.channelConfig = null;
+    Logger.fileChannel = null;
+    Logger.noticeCallback = null;
+  }
+
+  private isChannelEnabled(level: LogLevel, channel: "console" | "notice" | "file"): boolean {
+    if (!Logger.channelConfig) return channel === "console";
+    return Logger.channelConfig[level][channel];
+  }
+
+  private emitNotice(level: LogLevel, message: string): void {
+    if (!this.isChannelEnabled(level, "notice") || !Logger.noticeCallback) return;
+    const prefix = level === "error" ? "✗ " : level === "warn" ? "⚠ " : "";
+    Logger.noticeCallback(`${prefix}${message}`);
+  }
+
+  private emitFile(level: LogLevel, message: string): void {
+    if (!this.isChannelEnabled(level, "file") || !Logger.fileChannel) return;
+    Logger.fileChannel.append(level, this.context, message);
+  }
+
   debug(message: string, ...args: unknown[]): void {
-    console.debug(`[${this.context}] ${message}`, ...args);
+    if (this.isChannelEnabled("debug", "console")) {
+      console.debug(`[${this.context}] ${message}`, ...args);
+    }
+    this.emitNotice("debug", message);
+    this.emitFile("debug", message);
   }
 
   info(message: string, ...args: unknown[]): void {
-    console.info(`[${this.context}] ${message}`, ...args);
+    if (this.isChannelEnabled("info", "console")) {
+      console.info(`[${this.context}] ${message}`, ...args);
+    }
+    this.emitNotice("info", message);
+    this.emitFile("info", message);
   }
 
   warn(message: string, ...args: unknown[]): void {
-    console.warn(`[${this.context}] ${message}`, ...args);
+    if (this.isChannelEnabled("warn", "console")) {
+      console.warn(`[${this.context}] ${message}`, ...args);
+    }
+    this.emitNotice("warn", message);
+    this.emitFile("warn", message);
   }
 
   error(message: string, errorOrOptions?: Error | unknown | ErrorLogOptions): void {
@@ -76,6 +134,10 @@ export class Logger implements ILogger {
     } else {
       this.logSimpleError(message, errorOrOptions, isDev);
     }
+
+    // Notice and file channels always get the top-level message
+    this.emitNotice("error", message);
+    this.emitFile("error", message);
   }
 
   /**
@@ -104,6 +166,8 @@ export class Logger implements ILogger {
    * conditional logging into single console.error calls.
    */
   private logWithOptions(message: string, options: ErrorLogOptions, isDev: boolean): void {
+    if (!this.isChannelEnabled("error", "console")) return;
+
     const { errorCode, error, context } = options;
     const errorCodeStr = errorCode ? ` [${errorCode}]` : "";
 
@@ -157,6 +221,8 @@ export class Logger implements ILogger {
    * Note: Uses helper methods to avoid orphaned expressions after console drops.
    */
   private logSimpleError(message: string, error: unknown, isDev: boolean): void {
+    if (!this.isChannelEnabled("error", "console")) return;
+
     console.error(`[${this.context}] ${message}`);
 
     if (isDev) {

--- a/packages/obsidian-plugin/src/domain/settings/ExocortexSettings.ts
+++ b/packages/obsidian-plugin/src/domain/settings/ExocortexSettings.ts
@@ -33,6 +33,26 @@ export const DEFAULT_DISPLAY_NAME_SETTINGS: DisplayNameSettings = {
   },
 };
 
+/**
+ * Configuration for which channels a log level should route to.
+ */
+export interface LogChannelConfig {
+  notice: boolean;
+  console: boolean;
+  file: boolean;
+}
+
+export type LogLevel = "debug" | "info" | "warn" | "error";
+
+export type LogChannelsSettings = Record<LogLevel, LogChannelConfig>;
+
+export const DEFAULT_LOG_CHANNELS: LogChannelsSettings = {
+  debug: { notice: false, console: true, file: false },
+  info: { notice: false, console: true, file: false },
+  warn: { notice: true, console: true, file: false },
+  error: { notice: true, console: true, file: false },
+};
+
 export interface ExocortexSettings {
   layoutVisible: boolean;
   showArchivedAssets: boolean;
@@ -59,6 +79,8 @@ export interface ExocortexSettings {
    * @see Issue #2142
    */
   autoAdjustPlannedEndTimestamp: boolean;
+  /** Per-level log channel routing configuration */
+  logChannels: LogChannelsSettings;
   [key: string]: unknown;
 }
 
@@ -77,4 +99,5 @@ export const DEFAULT_SETTINGS: ExocortexSettings = {
   displayNameTemplate: "{{exo__Asset_label}} ({{exo__Instance_class}})",
   displayNameSettings: DEFAULT_DISPLAY_NAME_SETTINGS,
   autoAdjustPlannedEndTimestamp: false,
+  logChannels: DEFAULT_LOG_CHANNELS,
 };

--- a/packages/obsidian-plugin/src/presentation/settings/ExocortexSettingTab.ts
+++ b/packages/obsidian-plugin/src/presentation/settings/ExocortexSettingTab.ts
@@ -4,7 +4,9 @@ import { DEFAULT_DISPLAY_NAME_TEMPLATE } from "@plugin/domain/display-name/Displ
 import { DisplayNameResolver } from "@plugin/domain/display-name/DisplayNameResolver";
 import {
   DEFAULT_DISPLAY_NAME_SETTINGS,
+  DEFAULT_LOG_CHANNELS,
   type DisplayNameSettings,
+  type LogLevel,
 } from "@plugin/domain/settings/ExocortexSettings";
 
 export class ExocortexSettingTab extends PluginSettingTab {
@@ -139,6 +141,9 @@ export class ExocortexSettingTab extends PluginSettingTab {
           }),
       );
 
+    // Logging section
+    this.renderLogChannelsSection(containerEl);
+
     // Display Name Template section
     new Setting(containerEl)
       .setName("Display name templates")
@@ -248,6 +253,56 @@ export class ExocortexSettingTab extends PluginSettingTab {
       const li = placeholderList.createEl("li");
       li.createEl("code", { text: code });
       li.appendText(` - ${desc}`);
+    }
+  }
+
+  /**
+   * Render the log channel routing matrix.
+   * Rows = log levels, columns = channels (Notice / Console / File).
+   */
+  private renderLogChannelsSection(containerEl: HTMLElement): void {
+    new Setting(containerEl)
+      .setName("Log channels")
+      .setHeading();
+
+    const desc = containerEl.createDiv({ cls: "setting-item-description" });
+    desc.appendText(
+      "Choose which channels each log level should be routed to. " +
+      "File channel writes to exocortex-logs.txt in the vault root.",
+    );
+
+    // Ensure logChannels exists
+    if (!this.plugin.settings.logChannels) {
+      this.plugin.settings.logChannels = { ...DEFAULT_LOG_CHANNELS };
+    }
+
+    const levels: { key: LogLevel; label: string }[] = [
+      { key: "debug", label: "Debug" },
+      { key: "info", label: "Info" },
+      { key: "warn", label: "Warn" },
+      { key: "error", label: "Error" },
+    ];
+
+    const channels: { key: "notice" | "console" | "file"; label: string }[] = [
+      { key: "notice", label: "Notice" },
+      { key: "console", label: "Console" },
+      { key: "file", label: "File" },
+    ];
+
+    for (const level of levels) {
+      const setting = new Setting(containerEl).setName(level.label);
+
+      for (const channel of channels) {
+        setting.addToggle((toggle) =>
+          toggle
+            .setValue(this.plugin.settings.logChannels[level.key][channel.key])
+            .onChange(async (value) => {
+              this.plugin.settings.logChannels[level.key][channel.key] = value;
+              await this.plugin.saveSettings();
+              this.plugin.configureLogChannels();
+            }),
+        );
+      }
     }
   }
 

--- a/packages/obsidian-plugin/tests/unit/ExocortexSettingTab.test.ts
+++ b/packages/obsidian-plugin/tests/unit/ExocortexSettingTab.test.ts
@@ -55,11 +55,18 @@ describe("ExocortexSettingTab", () => {
             "ems__TaskPrototype": "{{exo__Asset_label}} (TaskPrototype)",
           },
         },
+        logChannels: {
+          debug: { notice: false, console: true, file: false },
+          info: { notice: false, console: true, file: false },
+          warn: { notice: true, console: true, file: false },
+          error: { notice: true, console: true, file: false },
+        },
       },
       saveSettings: jest.fn().mockResolvedValue(undefined),
       refreshLayout: jest.fn(),
       toggleTabTitleLabels: jest.fn(),
       applyDisplayNameTemplate: jest.fn(),
+      configureLogChannels: jest.fn(),
     });
 
     mockApp = createMockApp();
@@ -120,10 +127,9 @@ describe("ExocortexSettingTab", () => {
       settingTab.display();
 
       expect(mockContainerEl.empty).toHaveBeenCalled();
-      // 6 toggle settings + 2 headings + 1 default template + 6 per-class templates + 1 reset button = 16
-      // Removed: showPropertiesSection (Issue #2321), showLabelsInQuickSwitcher, showLabelsInWikilinkAutocomplete (Issue #2318)
-      // Removed: sortByDisplayName (Issue #2327)
-      expect(MockSetting).toHaveBeenCalledTimes(18);
+      // 9 toggle settings + 3 headings + 1 default template + 6 per-class templates + 1 reset button + 4 log level rows = 23
+      // Log channels section: 1 heading + 4 log level rows = 5
+      expect(MockSetting).toHaveBeenCalledTimes(23);
     });
 
     it("should render layout visibility toggle as first setting", () => {

--- a/packages/obsidian-plugin/tests/unit/FileLogChannel.test.ts
+++ b/packages/obsidian-plugin/tests/unit/FileLogChannel.test.ts
@@ -1,0 +1,109 @@
+import { FileLogChannel } from "../../src/adapters/logging/FileLogChannel";
+
+describe("FileLogChannel", () => {
+  let channel: FileLogChannel;
+  let mockAdapter: { append: jest.Mock; write: jest.Mock };
+
+  beforeEach(() => {
+    mockAdapter = {
+      append: jest.fn().mockResolvedValue(undefined),
+      write: jest.fn().mockResolvedValue(undefined),
+    };
+    channel = new FileLogChannel(mockAdapter as any);
+  });
+
+  it("should append log lines to the log file via adapter", async () => {
+    channel.append("info", "TestCtx", "Hello world");
+
+    // Wait for microtask flush
+    await flushMicrotasks();
+
+    expect(mockAdapter.append).toHaveBeenCalledTimes(1);
+    const writtenLine = mockAdapter.append.mock.calls[0][1] as string;
+    expect(writtenLine).toContain("[INFO]");
+    expect(writtenLine).toContain("[TestCtx]");
+    expect(writtenLine).toContain("Hello world");
+    expect(writtenLine).toMatch(/^\[\d{4}-\d{2}-\d{2}T/);
+    expect(writtenLine).toEndWith("\n");
+    expect(mockAdapter.append.mock.calls[0][0]).toBe("exocortex-logs.txt");
+  });
+
+  it("should batch multiple appends in the same tick", async () => {
+    channel.append("debug", "A", "Line 1");
+    channel.append("warn", "B", "Line 2");
+    channel.append("error", "C", "Line 3");
+
+    await flushMicrotasks();
+
+    expect(mockAdapter.append).toHaveBeenCalledTimes(1);
+    const combined = mockAdapter.append.mock.calls[0][1] as string;
+    expect(combined).toContain("[DEBUG]");
+    expect(combined).toContain("[WARN]");
+    expect(combined).toContain("[ERROR]");
+    expect(combined.split("\n").filter(Boolean)).toHaveLength(3);
+  });
+
+  it("should fallback to write when append fails (file does not exist)", async () => {
+    mockAdapter.append.mockRejectedValueOnce(new Error("ENOENT"));
+
+    channel.append("info", "Ctx", "First write");
+
+    await flushMicrotasks();
+    // Allow error handler promise to resolve
+    await flushMicrotasks();
+
+    expect(mockAdapter.write).toHaveBeenCalledTimes(1);
+    const writtenLine = mockAdapter.write.mock.calls[0][1] as string;
+    expect(writtenLine).toContain("First write");
+  });
+
+  it("should silently handle write failure", async () => {
+    mockAdapter.append.mockRejectedValueOnce(new Error("ENOENT"));
+    mockAdapter.write.mockRejectedValueOnce(new Error("EPERM"));
+
+    // Should not throw
+    channel.append("error", "Ctx", "Fail silently");
+
+    await flushMicrotasks();
+    await flushMicrotasks();
+
+    expect(mockAdapter.append).toHaveBeenCalledTimes(1);
+    expect(mockAdapter.write).toHaveBeenCalledTimes(1);
+  });
+
+  it("should format log lines with ISO timestamp", async () => {
+    channel.append("warn", "MyService", "Something happened");
+
+    await flushMicrotasks();
+
+    const line = mockAdapter.append.mock.calls[0][1] as string;
+    // ISO 8601 format check
+    const isoMatch = line.match(/^\[(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z)\]/);
+    expect(isoMatch).not.toBeNull();
+  });
+});
+
+function flushMicrotasks(): Promise<void> {
+  return new Promise(resolve => {
+    queueMicrotask(resolve);
+  });
+}
+
+// Custom matcher
+expect.extend({
+  toEndWith(received: string, suffix: string) {
+    const pass = received.endsWith(suffix);
+    return {
+      message: () => `expected "${received}" to end with "${suffix}"`,
+      pass,
+    };
+  },
+});
+
+declare global {
+  namespace jest {
+    interface Matchers<R> {
+      toEndWith(suffix: string): R;
+    }
+  }
+}

--- a/packages/obsidian-plugin/tests/unit/LogChannelSettings.test.ts
+++ b/packages/obsidian-plugin/tests/unit/LogChannelSettings.test.ts
@@ -1,0 +1,45 @@
+import {
+  DEFAULT_LOG_CHANNELS,
+  DEFAULT_SETTINGS,
+  type LogChannelsSettings,
+} from "../../src/domain/settings/ExocortexSettings";
+
+describe("LogChannelSettings", () => {
+  describe("DEFAULT_LOG_CHANNELS", () => {
+    it("should have all four log levels", () => {
+      expect(DEFAULT_LOG_CHANNELS).toHaveProperty("debug");
+      expect(DEFAULT_LOG_CHANNELS).toHaveProperty("info");
+      expect(DEFAULT_LOG_CHANNELS).toHaveProperty("warn");
+      expect(DEFAULT_LOG_CHANNELS).toHaveProperty("error");
+    });
+
+    it("should default debug to console only", () => {
+      expect(DEFAULT_LOG_CHANNELS.debug).toEqual({ notice: false, console: true, file: false });
+    });
+
+    it("should default info to console only", () => {
+      expect(DEFAULT_LOG_CHANNELS.info).toEqual({ notice: false, console: true, file: false });
+    });
+
+    it("should default warn to notice + console", () => {
+      expect(DEFAULT_LOG_CHANNELS.warn).toEqual({ notice: true, console: true, file: false });
+    });
+
+    it("should default error to notice + console", () => {
+      expect(DEFAULT_LOG_CHANNELS.error).toEqual({ notice: true, console: true, file: false });
+    });
+
+    it("should not enable file channel by default for any level", () => {
+      const levels = Object.keys(DEFAULT_LOG_CHANNELS) as (keyof LogChannelsSettings)[];
+      for (const level of levels) {
+        expect(DEFAULT_LOG_CHANNELS[level].file).toBe(false);
+      }
+    });
+  });
+
+  describe("DEFAULT_SETTINGS", () => {
+    it("should include logChannels with default values", () => {
+      expect(DEFAULT_SETTINGS.logChannels).toEqual(DEFAULT_LOG_CHANNELS);
+    });
+  });
+});

--- a/packages/obsidian-plugin/tests/unit/Logger.test.ts
+++ b/packages/obsidian-plugin/tests/unit/Logger.test.ts
@@ -24,6 +24,7 @@ describe("Logger", () => {
   });
 
   afterEach(() => {
+    Logger.resetChannels();
     jest.restoreAllMocks();
   });
 
@@ -336,6 +337,156 @@ describe("Logger", () => {
       longLogger.info("test");
       expect(consoleInfoSpy).toHaveBeenCalledWith(`[${longContext}] test`);
     });
+  });
+});
+
+describe("Logger channel routing", () => {
+  let logger: Logger;
+  let consoleDebugSpy: jest.SpyInstance;
+  let consoleInfoSpy: jest.SpyInstance;
+  let consoleWarnSpy: jest.SpyInstance;
+  let consoleErrorSpy: jest.SpyInstance;
+
+  beforeEach(() => {
+    consoleDebugSpy = jest.spyOn(console, "debug").mockImplementation();
+    consoleInfoSpy = jest.spyOn(console, "info").mockImplementation();
+    consoleWarnSpy = jest.spyOn(console, "warn").mockImplementation();
+    consoleErrorSpy = jest.spyOn(console, "error").mockImplementation();
+    Logger.setDevelopmentMode(false);
+    logger = new Logger("ChannelTest");
+  });
+
+  afterEach(() => {
+    Logger.resetChannels();
+    jest.restoreAllMocks();
+  });
+
+  it("should route to console by default when no config set", () => {
+    logger.info("default message");
+    expect(consoleInfoSpy).toHaveBeenCalledWith("[ChannelTest] default message");
+  });
+
+  it("should suppress console when console channel is disabled", () => {
+    Logger.configure({
+      channels: {
+        debug: { notice: false, console: false, file: false },
+        info: { notice: false, console: false, file: false },
+        warn: { notice: false, console: false, file: false },
+        error: { notice: false, console: false, file: false },
+      },
+    });
+
+    logger.debug("hidden");
+    logger.info("hidden");
+    logger.warn("hidden");
+    logger.error("hidden");
+
+    expect(consoleDebugSpy).not.toHaveBeenCalled();
+    expect(consoleInfoSpy).not.toHaveBeenCalled();
+    expect(consoleWarnSpy).not.toHaveBeenCalled();
+    expect(consoleErrorSpy).not.toHaveBeenCalled();
+  });
+
+  it("should route to notice callback when notice channel is enabled", () => {
+    const noticeFn = jest.fn();
+    Logger.configure({
+      channels: {
+        debug: { notice: false, console: false, file: false },
+        info: { notice: true, console: false, file: false },
+        warn: { notice: true, console: false, file: false },
+        error: { notice: true, console: false, file: false },
+      },
+      noticeCallback: noticeFn,
+    });
+
+    logger.info("visible notice");
+    expect(noticeFn).toHaveBeenCalledWith("visible notice");
+
+    logger.warn("warn notice");
+    expect(noticeFn).toHaveBeenCalledWith("⚠ warn notice");
+
+    logger.error("error notice");
+    expect(noticeFn).toHaveBeenCalledWith("✗ error notice");
+  });
+
+  it("should not emit notice when notice channel is disabled", () => {
+    const noticeFn = jest.fn();
+    Logger.configure({
+      channels: {
+        debug: { notice: false, console: true, file: false },
+        info: { notice: false, console: true, file: false },
+        warn: { notice: false, console: true, file: false },
+        error: { notice: false, console: true, file: false },
+      },
+      noticeCallback: noticeFn,
+    });
+
+    logger.info("no notice");
+    logger.warn("no notice");
+    logger.error("no notice");
+
+    expect(noticeFn).not.toHaveBeenCalled();
+  });
+
+  it("should route to file channel when file channel is enabled", () => {
+    const fileChannel = { append: jest.fn() } as any;
+    Logger.configure({
+      channels: {
+        debug: { notice: false, console: false, file: true },
+        info: { notice: false, console: false, file: true },
+        warn: { notice: false, console: false, file: false },
+        error: { notice: false, console: false, file: false },
+      },
+      fileChannel,
+    });
+
+    logger.debug("file debug");
+    expect(fileChannel.append).toHaveBeenCalledWith("debug", "ChannelTest", "file debug");
+
+    logger.info("file info");
+    expect(fileChannel.append).toHaveBeenCalledWith("info", "ChannelTest", "file info");
+
+    logger.warn("no file warn");
+    expect(fileChannel.append).toHaveBeenCalledTimes(2);
+  });
+
+  it("should support multiple channels simultaneously", () => {
+    const noticeFn = jest.fn();
+    const fileChannel = { append: jest.fn() } as any;
+    Logger.configure({
+      channels: {
+        debug: { notice: false, console: true, file: false },
+        info: { notice: false, console: true, file: false },
+        warn: { notice: true, console: true, file: true },
+        error: { notice: true, console: true, file: true },
+      },
+      noticeCallback: noticeFn,
+      fileChannel,
+    });
+
+    logger.warn("all channels");
+    expect(consoleWarnSpy).toHaveBeenCalledWith("[ChannelTest] all channels");
+    expect(noticeFn).toHaveBeenCalledWith("⚠ all channels");
+    expect(fileChannel.append).toHaveBeenCalledWith("warn", "ChannelTest", "all channels");
+  });
+
+  it("should reset channels correctly", () => {
+    Logger.configure({
+      channels: {
+        debug: { notice: false, console: false, file: false },
+        info: { notice: false, console: false, file: false },
+        warn: { notice: false, console: false, file: false },
+        error: { notice: false, console: false, file: false },
+      },
+    });
+
+    logger.info("suppressed");
+    expect(consoleInfoSpy).not.toHaveBeenCalled();
+
+    Logger.resetChannels();
+
+    logger.info("back to default");
+    expect(consoleInfoSpy).toHaveBeenCalledWith("[ChannelTest] back to default");
   });
 });
 

